### PR TITLE
Added support for translations

### PR DESCRIPTION
**Pull request by christianlupus**
Worked along https://docs.nextcloud.com/server/stable/developer_manual/app/view/l10n.html with manual process. If @mrzapp  wants to use transifexit might be better if you do it.

This is related to issue #21.

I did not write any documentation how to generate the translation files. Also I did not check in the generated pot and js files into git. You might want to define a best practice here.

### CODE REVIEW

File: translationfiles/de/cookbook.po [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+msgid "Are you sure you want to delete this recipe?"
+msgstr "Soll das Rezept auch wirklich gelöscht werden?"
+
+#: /srv/http/nc/apps/cookbook/js/script.js:331
+msgid "Failed to delete recipe"
+msgstr "Löschen des Rezepts ist fehlgeschlagen"
+
+#: /srv/http/nc/apps/cookbook/js/script.js:362
+msgid "Failed to fetch recipes"
+msgstr "Laden des Rezepts ist fehlgeschlagen"
```
Comment (clesmian) [inline translationfiles/de/cookbook.po]:
```suggestion
msgstr "Laden der Rezepte ist fehlgeschlagen"
```

File: translationfiles/de/cookbook.po [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+msgid "Could not add recipe"
+msgstr "Konnte Rezept nicht hinzufügen"
+
+#: /srv/http/nc/apps/cookbook/js/script.js:300
+msgid "Could not rebuild recipe index."
+msgstr "Konnte den Rezept-Index nicht neu aufbauen."
+
+#: /srv/http/nc/apps/cookbook/js/script.js:315
+msgid "Are you sure you want to delete this recipe?"
+msgstr "Soll das Rezept auch wirklich gelöscht werden?"
```
Comment (clesmian) [inline translationfiles/de/cookbook.po]:
```suggestion
msgstr "Soll das Rezept wirklich gelöscht werden?"
```

File: translationfiles/de/cookbook.po [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+"Plural-Forms: nplurals=2; plural=(n != 1);\n"
+"X-Generator: Lokalize 19.04.3\n"
+
+#: /srv/http/nc/apps/cookbook/js/script.js:91
+msgid "Path to your recipe collection"
+msgstr "Pfad zur Rezept-Datenbank"
+
+#: /srv/http/nc/apps/cookbook/js/script.js:106
+msgid "Could not set recipe folder to {path}"
+msgstr "Konnte den Pfad zu den Rezepten auf {path} setzen"
```
Comment (clesmian) [inline translationfiles/de/cookbook.po]:
```suggestion
msgstr "Konnte den Pfad '{path}' nicht als Rezept-Ordner setzen"
```

File: translationfiles/de/cookbook.po [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+"Language: de_DE\n"
+"MIME-Version: 1.0\n"
+"Content-Type: text/plain; charset=UTF-8\n"
+"Content-Transfer-Encoding: 8bit\n"
+"Plural-Forms: nplurals=2; plural=(n != 1);\n"
+"X-Generator: Lokalize 19.04.3\n"
+
+#: /srv/http/nc/apps/cookbook/js/script.js:91
+msgid "Path to your recipe collection"
+msgstr "Pfad zur Rezept-Datenbank"
```
Comment (clesmian) [inline translationfiles/de/cookbook.po]:
```suggestion
msgstr "Pfad zur Rezept-Sammlung"
```

File: translationfiles/de/cookbook.po [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+msgid ""
+"A library for all your recipes. It uses JSON files following the schema.org "
+"recipe format. To add a recipe to the collection, you can paste in the URL "
+"of the recipe, and the provided web page will be parsed and downloaded to "
+"whichever folder you specify in the app settings."
+msgstr ""
+"Eine Bibliothek aller Rezepte. Sie nutzt JSON Dateien, die dem schema.org"
+" Standard für Rezepte folgen. "
+"Um weitere Rezepte zu der Sammlung hinzuzufügen, kann die URL des Rezepts"
+" eingefügt werden, so dass die Seite heruntergeladen, analysiert und in einem"
```
Comment (clesmian) [inline translationfiles/de/cookbook.po]:
```suggestion
" eingefügt werden, so dass die Seite heruntergeladen, verarbeitet und in einem"
```

File: translationfiles/de/cookbook.po [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+msgid "Name"
+msgstr "Name"
+
+#: /srv/http/nc/apps/cookbook/templates/content/edit.php:8
+msgid "Image URL"
+msgstr "Bild-URL"
+
+#: /srv/http/nc/apps/cookbook/templates/content/edit.php:13
+msgid "Keywords (comma-separated)"
+msgstr "Schlagworte (mit Kommata abgetrennt)"
```
Comment (clesmian) [inline translationfiles/de/cookbook.po]:
```suggestion
msgstr "Schlagworte (durch Kommas getrennt)"
```

File: translationfiles/de/cookbook.po [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+
+#: /srv/http/nc/apps/cookbook/templates/content/edit.php:26
+#: /srv/http/nc/apps/cookbook/templates/content/recipe.php:18
+msgid "Berries"
+msgstr "Beeren"
+
+#: /srv/http/nc/apps/cookbook/templates/content/edit.php:27
+#: /srv/http/nc/apps/cookbook/templates/content/recipe.php:19
+msgid "Cruciferous vegetables"
+msgstr "Kreuzblütler-Gemüse"
```
Comment (clesmian) [inline translationfiles/de/cookbook.po]:
```suggestion
msgstr "Kreuzblütler-Gemüse (z.B. Kohl, Brokkoli oder Radieschen)"
```
Could be a little bit clearer to people that don't know the term.

File: translationfiles/de/cookbook.po [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+#: /srv/http/nc/apps/cookbook/templates/content/edit.php:28
+#: /srv/http/nc/apps/cookbook/templates/content/recipe.php:20
+msgid "Flaxseeds"
+msgstr "Leinsamen"
+
+#: /srv/http/nc/apps/cookbook/templates/content/edit.php:29
+#, fuzzy
+#: /srv/http/nc/apps/cookbook/templates/content/recipe.php:21
+msgid "Greens"
+msgstr ""
```
Comment (clesmian) [inline translationfiles/de/cookbook.po]:
```suggestion
msgstr "Blattgemüse"
```

**Comment by mrzapp** [comment]
@christianlupus Thanks for your contribution! Are the compiled translation files automatically generated without user interaction? If not, then I'd suggest adding them to git.

**Comment by christianlupus** [comment]
> @christianlupus Thanks for your contribution! Are the compiled translation files automatically generated without user interaction? If not, then I'd suggest adding them to git.

No, they are not automatically generated. To be honest, this was a big question mark on my side how you deploy the app. This is something that should be done in CD (continuous deplyment). Of course I can add them to git but this is not best git practice as you know. Therefore I asked you for your preferences.

File: translationfiles/de/cookbook.po (hunk 6 above)
Comment (christianlupus) [reply translationfiles/de/cookbook.po]: @clesmian The Plural of Komma is Kommata, right? So it would be `durch Kommata getrennt`.

**Comment by mrzapp** [comment]
@christianlupus yeah, I'm aware. I don't have CD set up for this project, I honestly didn't expect it to gain so much traction. I can also add it to my build script, if you let me know how you compile them?

**Comment by christianlupus** [comment]
Sure, I send you the condensed details:

- Clone/update the [translation repo](https://github.com/nextcloud/docker-ci/tree/master/translations/translationtool)
- First step is to go to the roo folder of your app and call `php /path/tp/translationtool.phar create-pot-files`. This will generate the template for new translations. Might come in handy if you extend the app later to allow the translators update the language files.
- To create a new language just copy the template `cookbook.pot` to the corresponding `po` file (like I did with the German translation for testing).
- To generate the files nextcluoud needs in order to render the page in localized way, ensure the folder `l10n` exists and call `php /path/tp/translationtool.phar convert-po-files`. 

This last step is the only thing needed for CD the rest is only for development/localization.

File: translationfiles/de/cookbook.po (hunk 6 above)
Comment (clesmian) [reply translationfiles/de/cookbook.po]: According to duden.de both is correct. I guess "kommagetrennt" would also be correct and probably closer to the original.

To be honest I don't know which version is best.

**Pull request merged**